File: ml-class-ex1/ex1/Python-Version/ex1.py
#!/usr/bin/env python
# coding: utf-8

# In[45]:


import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt

path = os.getcwd()+"/Documents/ML Coursera/ml-class-ex1/ex1/Python-Version/"
os.getcwd()

#data = open(os.listdir(os.getcwd())[0]).read()
column_names = ['population','profit']
data = pd.read_csv('ex1data1.txt', sep=",", header=None, names = column_names)
#data.iloc[0:1,0:3]
#data.plot.scatter(x='population',
#                  y='profit',
#                  c='red')

#Add a column of ones to x
data['x0'] = 1
X = data.iloc[:,[2,0]]
y = np.array(data.iloc[:,1])
theta = np.array([0,0])
iterations = 1500
alpha = 0.01

def costFunction(X, y, theta):
    m = len(X)
    predictions = np.dot(X,theta)
    sqrErrors = np.array((predictions-y))**2
    J = 1/(2*m)*np.sum(sqrErrors)
    return J;


#def gradientDescent(X, y, theta, alpha, num_iter):
#    m= len(X)
#    for i in range(0, num_iter):
#        temp0 = 


# In[46]:


def gradientDescent(X, y, theta, alpha, num_iter):
    m= len(X)
    for i in range(0, num_iter):
        temp0 = theta[0]-(alpha/m)*sum((np.dot(X,theta)-y)*X.iloc[:,0])
        temp1 = theta[1]-(alpha/m)*sum((np.dot(X,theta)-y)*X.iloc[:,1])
        theta = np.array([temp0,temp1])
        costFunction(X, y, theta)
    print(costFunction(X, y, theta))
    


# In[47]:


gradientDescent(X,y,theta,0.01, 100)

File: ml-class-ex1/R version/ex1.R
##Ex 1 ML Coursera##
library(tidyverse)
library(ggplot2)

data <- read.delim("ex1data1.txt", header = F, sep = ",")
data%>%ggplot(aes(x = V1, y = V2))+
  geom_point(shape = 4, color = "red")+
  labs(x = "Population of Cityin 10,000s", y = "Profit in $10,000s")
  
## Getting proper X matrix
colnames(data)<-c("population", "profit")
X = cbind(data, as.vector(matrix(1,1, length(data$population))))
X = cbind(X[,3],X[,1])
colnames(X)<- c("x0", "population")
head(X,2)
#Initializating rest of variables
y = as.vector(data[,2])
theta<-as.vector(rbind(0,0))



costFunction<- function(X, y, theta){
  m <- nrow(X)
  predictions <- X%*%theta
  sqrErrors <- (predictions-y)^2
  J <- 1/(2*m)*sum(sqrErrors)
  J
}

costFunction(X, y, theta)

gradientDescent<-function(X, y, theta, alpha, num_iter){
  m <- nrow(X)
  for(i in 1:num_iter){
    temp0<-theta[1]-(alpha/m)*sum((X%*%theta-y)*X[,1])
    temp1<-theta[2]-(alpha/m)*sum((X%*%theta-y)*X[,2])
    theta<-c(temp0,temp1)
    costFunction(X, y, theta)
    
  }
  costFunction(X, y, theta)
  theta
}
theta<-as.vector(rbind(0,0))

gradientDescent(X,y,theta,0.01, 100)
